tests/unit/core/testWithUIState.js

```diff
@@ -0,0 +1,160 @@
+import { setUIState } from 'core/reducers/uiState';
+import { dispatchClientMetadata } from 'tests/unit/amo/helpers';
+import { applyUIStateChanges, shallowUntilTarget } from 'tests/unit/helpers';
+
+describe(__filename, () => {
+  describe('withUIState', () => {
+    let store;
+
+    class OverlayBase extends React.Component {
+      closeOverlay = () => {
+        this.props.setUIState({ isOpen: false });
+      };
+
+      render() {
+        const { uiState } = this.props;
+
+        if (!uiState.isOpen) {
+          return null;
+        }
+
+        return (
+          <div className="overlay">
+            <button className="close-button" onClick={this.closeOverlay}>
+              Close overlay
+            </button>
+          </div>
+        );
+      }
+    }
+
+    const Overlay = compose(
+      withUIState({
+        fileName: __filename,
+        extractId: (props) => props.id,
+        initialState: { isOpen: true },
+      }),
+    )(OverlayBase);
+
+    const render = ({ id = 'some-id', ...props } = {}) => {
+      const root = shallowUntilTarget(
+        <Overlay id={id} store={store} {...props} />,
+        OverlayBase,
+      );
+      // Apply initial state change from componentWillMount.
+      applyUIStateChanges({ root, store });
+      return root;
+    };
+
+    beforeEach(() => {
+      store = dispatchClientMetadata().store;
+    });
+
+    it('lets you manage UI state', () => {
+      const root = render();
+      expect(root.find('.overlay')).toHaveLength(1);
+
+      // Test that closing the overlay is hooked up to uiState.
+      root.find('.close-button').simulate('click');
+      applyUIStateChanges({ root, store });
+
+      expect(root.find('.overlay')).toHaveLength(0);
+    });
+
+    it('separates state by instance', () => {
+      const root1 = render({ id: 'one' });
+      const root2 = render({ id: 'two' });
+
+      // The first overlay should be open.
+      expect(root1.find('.overlay')).toHaveLength(1);
+
+      // Close the second overlay.
+      root2.find('.close-button').simulate('click');
+
+      applyUIStateChanges({ root: root1, store });
+      applyUIStateChanges({ root: root2, store });
+
+      // The first overlay should still be open.
+      expect(root1.find('.overlay')).toHaveLength(1);
+    });
+
+    it('begins with an initial state', () => {
+      class ThingBase extends React.Component {
+        render() {
+          return <div />;
+        }
+      }
+      const initialState = { visible: true };
+
+      const Thing = compose(
+        withUIState({
+          fileName: __filename,
+          extractId: () => '',
+          initialState,
+        }),
+      )(ThingBase);
+
+      const root = shallowUntilTarget(<Thing store={store} />, ThingBase);
+
+      expect(root.instance().props.uiState).toEqual(initialState);
+    });
+
+    it('always resets initial state per component instance', () => {
+      class ThingBase extends React.Component {
```

Comment:
I think you could probably use a functional component to reduce the amount of code in this test:

```
const ThingBase = () => <div />;
```

Reply:
You can't use `root.instance().props` on a stateless component because there is no instance. You also can't use `root.props()` because that only returns *rendered* props, not props passed to the stateless component's function. I think `instance()` is helpful here because we are testing the affect of *injecting* props into a component. It's not testing the rendering of a component, unlike our typical usage of Enzyme.

Reply:
ah, I see. But the HOC should work with stateless components though, no?

Reply:
Yes, the HOC will work fine with stateless components. The `root.instance().props` call is only necessary to make an assertion about the received props while testing.